### DISCUSSION TITLE
Building an SDK with Viem

### DISCUSSION THREAD

**Discussion by gershido** [wevm/viem#831] [Question]
Hey :) Question about how to handle `PublicClient` or `WalletClient` types when building an SDK which uses Viem. 
The SDK client is initialized by the user providing a `PublicClient` and `WalletClient`. This is a common practice for SDKs, for example, SDKs that use Ethers are usually initialized by the user providing a `Provider` or `Signer`. 
Currently I require the user to use an exact version of Viem, otherwise, if the user uses a different patch version then the types are non compatible. I wonder if there is a better way, as this is not a very friendly requirement. 
Thank you!

**Comment by jxom** [comment]
Could you make `viem` a peer dependency so you don't have to keep the consumer version in sync?

**Reply by gershido** [reply]
That's my current approach. I'm using Viem 1.2.6 and having that also as a peer dependency. But that means that I require the consumer to use exactly this version or have 2 different versions installed if he/she wants a to use a different Viem version in their app. I wondered if there's maybe a base type that we can use that stays the same between versions, or any typescript trick I'm not familiar with.

**Comment by chybisov** [comment]
We are considering moving our SDK from ethers to viem for the next major, and I just discovered this discussion. 

@gershido could you please share which versions of viem have different types? It's okay if that's a major 0.x.x vs 1.x.x, but it would be concerning if that's some minor changes of such an important type between minor/patch versions. 

Thank you!

**Reply by gershido** [reply]
Hey, the docs mention this explicitly [here](https://viem.sh/docs/typescript.html#:~:text=Changes%20to%20types%20in%20this%20repository%20are%20considered%20non%2Dbreaking%20and%20are%20usually%20released%20as%20patch%20semver%20changes%20(otherwise%20every%20type%20enhancement%20would%20be%20a%20major%20version!).)

I noticed the non matching types between 1.2.6 and 1.2.7 I think.

**Reply by chybisov** [reply]
Wow, that's unexpected. Thank you for pointing this out. 

@jxom is there any chance to change this strategy as this is not a friendly behavior for library authors who want to have `viem` as a dependency? At least it would be awesome to stabilize PublicClient/WalletClient types so they can be safely used in a way described by @gershido. Thank you!
